src/System.Text.Encodings.Web/src/System/Text/Encodings/Web/TextEncoder.cs

```diff
@@ -813,10 +904,26 @@ private byte[] GetAsciiEncoding(byte value)
                     _asciiEscape[value] = encoding;
                 }
             }
-
             return encoding;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int DoesAsciiNeedEncoding(byte value)
+        {
+            Debug.Assert(value <= 0x7F);
+
+            int needsEscaping = _asciiNeedsEscaping[value];
+
+            Debug.Assert(needsEscaping == 0 || needsEscaping == 1 || needsEscaping == -1);
+
+            if (needsEscaping == 0)
+            {
+                needsEscaping = WillEncode(value) ? 1 : -1;
+                _asciiNeedsEscaping[value] = needsEscaping;
```

Comment:
`DoesAsciiNeedEncoding` is used quite often, so this method should be as tiny as possible.
Therefore move this lazy assignement of the values to the ctor (`WillEncode` is already known there)?
Then this method is just the array-lookup.

Reply:
Which ctor? `TextEncoder` is an abstract class.

Reply:
> Which ctor?

Currently there is none, but you can create a protected ctor?

Reply:
The types that implement `TextEncoder` are also what implement `WillEncode`. If we added a protected ctor to `TextEncoder` it will end up calling `WillEncode` on the concrete types before the concrete type's ctor finishes.

So:
`DefaultJavaScriptEncoder` ctor starts -> Calls base `TextEncoder` ctor -> Calls `DefaultJavaScriptEncoder.WillEncode` -> This fails because the `DefaultJavaScriptEncoder Ctor` isn't complete yet (including setting up `_allowedCharacters`).

We could, theoretically, copy/paste and override the base-types `FindFirstCharacterToEncodeUtf8` for `DefaultJavaScriptEncoder` and duplicate that logic.

Reply:
Or call an initialize-method (outside of the loop which uses `DoesAsciiNeedEncoding`) that ensures the array is populated with data?

Reply:
I think the ~80% case has been addressed already (with the built-in encoders). I am less inclined to optimize `TextEncoder` itself any further since most people wouldn't be calling those methods anyway (at least when it comes to `S.T.Json`).

We can do that separately, for 5.0/master (I'd like to port this PR to 3.1). Looks like we can get ~15% perf win from doing so.

```C#
private bool _asciiCacheNeedsInit = true;

public virtual unsafe int FindFirstCharacterToEncodeUtf8(ReadOnlySpan<byte> utf8Text)
{
    if (_asciiCacheNeedsInit)
    {
        InitializeAsciiCache();
    }
    // ... rest of the current impl
}

[MethodImpl(MethodImplOptions.NoInlining)]
private void InitializeAsciiCache()
{
    Debug.Assert(_asciiCacheNeedsInit);

    for (int i = 0; i < _asciiNeedsEscaping.Length; i++)
    {
        _asciiNeedsEscaping[i] = WillEncode(i) ? 1 : -1;
    }
    _asciiCacheNeedsInit = false;
}
```

Reply:
I touched this in https://github.com/dotnet/corefx/pull/42073 also.